src/functions.cpp

```diff
@@ -0,0 +1,204 @@
+#include "functions.h"
+
+#include "log.h"
+
+namespace bpftrace {
+
+namespace {
+std::string arg_types_str(const std::vector<SizedType> &arg_types)
+{
+  std::string str = "(";
+  bool first = true;
+  for (const SizedType &arg_type : arg_types) {
+    if (!first)
+      str += ", ";
+    str += typestr(arg_type);
+    first = false;
+  }
+  str += ")";
+  return str;
+}
+
+std::string param_types_str(const std::vector<Param> &params)
+{
+  std::string str = "(";
+  bool first = true;
+  for (const Param &param : params) {
+    if (!first)
+      str += ", ";
+    str += typestr(param.type());
+    first = false;
+  }
+  str += ")";
+  return str;
+}
+} // namespace
+
+FunctionRegistry::FunctionRegistry(std::ostream &out) : out_(out)
+{
+}
+
+Function &FunctionRegistry::add(Function::Origin origin,
+                                std::string_view name,
+                                const SizedType &returnType,
+                                std::vector<Param> params)
+{
+  return add(origin, {}, name, returnType, params);
+}
+
+Function &FunctionRegistry::add(Function::Origin origin,
+                                std::string_view ns,
+                                std::string_view name,
+                                const SizedType &returnType,
+                                std::vector<Param> params)
+{
+  all_funcs_.push_back(std::make_unique<Function>(
+      origin, std::string{ name }, returnType, params));
+  Function &newFunc = *all_funcs_.back().get();
+  std::string fq_name;
+  if (ns.empty())
+    fq_name = name;
+  else
+    fq_name = std::string{ ns } + "::" + newFunc.name();
```

Comment:
Not a huge fan of using strings with implicit structure. Probably better to have a struct that can be used as a key (hashable and equality methods). All the colon stuff in probe attachpoints got really messy and hard to understand after a while.